pkg/auto/history/factory.go

```diff
@@ -157,26 +158,56 @@ func (a *automation) applyConfig(ctx context.Context, cfg config.Root) error {
 	// work out where we're getting the records from
 	var serverClient wrap.ServiceUnwrapper
 	payloads := make(chan []byte)
-	var collect func(context.Context, config.Source, chan<- []byte)
+	var collect collector
 	switch cfg.Source.Trait {
 	case trait.AirQualitySensor:
 		serverClient = gen.WrapAirQualitySensorHistory(historypb.NewAirQualitySensorServer(store))
-		collect = a.collectAirQualityChanges
+
+		if cfg.Source.Sample != nil && !cfg.Source.Sample.Disabled {
+			collect = a.sampleAirQualityChanges
+		} else {
+			collect = a.collectAirQualityChanges
+		}
```

Comment:
I don't think this logic belongs here. The collect function is already given the config.Source so can make up its own mind about how to collect things. Maybe as part of the de-dupe change this would become clearer. I'm thinking a package local util func like

```go
func collectChanges(ctx context.Context, cfg config.Source, pullFn, pollFn pull.GetFunc, changes chan-> []byte, logger *zap.Logger) error {
  // if cfg is using schedule
  //    setup poll loop calling pollFn
  // else
  //    call poll.Changes
}
```

Reply:
Doesn't have to be nil, could be a function that returns `status.Error(codes.Unsupported, "poll only")`, or something like that. Though I don't think this is _really_ what we want as we'd be using the pull.Changes poll behaviour instead of being in control of it here.